source/rfc/metrics.rst

```diff
@@ -0,0 +1,91 @@
+Metrics
+=======
+
+Problem
+-------
+
+Zuul jobs are run as builds on independent, created on-demand, nodes. No information is available about the
+state changes of the nodes throughout the whole buildset run in terms of CPU, memory, storage usage
+or any other useful information related to the buildset itself for that matter. The only results coming
+out of buildsets currently are test reports, run logs and built artifacts.
+
+To have better understanding and control over the CI system and any insight on how changes in code
+impact the building and testing processes, more information about the system the buildsets are being executed on
+needs to be gathered and on the outputs of the builds themselves.
+
+Metrics
+-------
+
+This section contains proposals of metrics to gather during and from buildsets.
+
+#. General
+
+  * run duration
+
+#. Physical resources
+
+  * CPU usage by time
+  * memory usage by time
+  * storage usage by time
+  * disk IO by time
+  * network IO by time
+  * CPU hours
+  * peak memory usage
+  * peak disk usage
+
+  *Note: in case of storage usage, might need to consider only specific mount points*
+
+#. Test jobs specific metrics
+
+  * number of tests run for a specific buildset - might be useful in enforcing that more tests are added
+  * duration of each test case execution
+  * run log size
+
+#. Packaging jobs specific metrics
+
+  * size of created packages
+
+#. Container builder jobs specific metrics
+
+  * size of created images
+
+Technical requirements
+----------------------
+
+#.  The metric gathering system should allow for storing two types of metrics:
+
+  * time-series data (e.g. CPU usage at given point in time)
+  * discrete values (e.g. built packages size)
+
+#.  Each buildsets metrics should be gathered separately, with consideration of review and patchset ID's. This
+    would require for the metrics naming convention to be of the form:
+
+      buildset.<pipeline>.<project-name>.<job-name>.<gerrit-change-id>.<gerrit-patchset>.<resource>
+
+    for the check, gate and experimental pipelines. For the other pipelines the git branch and a timestamp of the start
+    of the jobs should be used instead of gerrit-change-id and gerrit-patchset:
+
+      buildset.<pipeline>.<project-name>.<job-name>.<branch>.<job-start-timestamp>.<resource>
```

Comment:
Could we use variables such as `<build_number>` or `<zuul.buildset>` instead of `<job-start-timestamp>`? I feel like timestamp, while unique, is not commonly known and not too easy to find, compared to the other two.

Reply:
variable examples: http://logs.opencontrail.org/periodic-nightly/review.opencontrail.org/master/111/contrail-vnc-build-containers-centos7-ocata/zuul-info/inventory.yaml